python/lsst/pipe/base/graphBuilder.py

```diff
@@ -191,22 +126,143 @@ def dimensions(self) -> DimensionGraph:
         base = self.universe.empty
         if len(self) == 0:
             return base
-        return base.union(*[scaffolding.dimensions for scaffolding in self.values()])
+        return base.union(*[datasetType.dimensions for datasetType in self.keys()])
+
+    def unpackSingleRefs(self) -> NamedKeyDict[DatasetType, DatasetRef]:
+        """Unpack nested single-element `DatasetRef` dicts into a new
+        mapping with `DatasetType` keys and `DatasetRef` values.
 
-    def unpackRefs(self) -> NamedKeyDict:
-        """Unpack nested single-element `DatasetRef` lists into a new
-        dictionary.
+        This method assumes that each nest contains exactly one item, as is the
+        case for all "init" datasets.
 
-        This method assumes that each `_DatasetScaffolding.refs` list contains
-        exactly one `DatasetRef`, as is the case for all "init" datasets.
+        Returns
+        -------
+        dictionary : `NamedKeyDict`
+            Dictionary mapping `DatasetType` to `DatasetRef`, with both
+            `DatasetType` instances and string names usable as keys.
+        """
+        def getOne(refs: Dict[DataCoordinate, DatasetRef]) -> DatasetRef:
+            ref, = refs.values()
+            return ref
+        return NamedKeyDict({datasetType: getOne(refs) for datasetType, refs in self.items()})
+
+    def unpackMultiRefs(self) -> NamedKeyDict[DatasetType, DatasetRef]:
+        """Unpack nested multi-element `DatasetRef` dicts into a new
+        mapping with `DatasetType` keys and `set` of `DatasetRef` values.
 
         Returns
         -------
         dictionary : `NamedKeyDict`
             Dictionary mapping `DatasetType` to `DatasetRef`, with both
             `DatasetType` instances and string names usable as keys.
         """
-        return NamedKeyDict((datasetType, scaffolding.refs[0]) for datasetType, scaffolding in self.items())
+        return NamedKeyDict({datasetType: list(refs.values()) for datasetType, refs in self.items()})
+
+    def extract(self, datasetType: DatasetType, dataIds: Iterable[DataCoordinate]
+                ) -> Iterator[DatasetRef]:
+        """Iterate over the contained `DatasetRef` instances that match the
+        given `DatasetType` and data IDs.
+
+        Parameters
+        ----------
+        datasetType : `DatasetType`
+            Dataset type to match.
+        dataIds : `Iterable` [ `DataCoordinate` ]
+            Data IDs to match.
+
+        Yields
```

Comment:
Agreed, in the sense that this is what the Python community expects, but I'm curious as to whether you know if this actually matters.  AFAICT, whether an iterator-returning function is implemented as a generator or not is completely transparent to the caller, so distinguishing between them is always _just_ exposing an implementation detail.  Is that wrong in some edge cases (or maybe coroutines)?

Reply:
A few thoughts:
* pedantically the function does not yield anything it returns something that yields something. Its like the difference between iterable and iterator
* If someone really wanted to know the call stack (or something like that), this will never be re-entered, the thing you are returning will be
* It does make a difference when coroutines are involved (in so far as you loose the richer type information provided to any checkers)

I think in day to day practice you can use this as is, but your terminology does not really match what is happening, and if it was used like this in other places (co routines) it would be wrong. Code in a generator function is not executed until the first use (everything up to the first yield, which is why it must be "primed" with a next, but in this case there is code execution each time before the generator is returned. If someone really cared about that for some reason then the info you wrote would lead them astray. In practice it probably doesnt matter much.